pkg/ingester/activeseries/active_series.go

```diff
@@ -200,34 +225,77 @@ func (s *seriesStripe) updateSeriesTimestamp(now time.Time, series labels.Labels
 	}
 }
 
-func (s *seriesStripe) findEntryForSeries(ref uint64) *atomic.Int64 {
+func (s *seriesStripe) findEntryForSeries(ref uint64, hasNativeHistograms bool, numNativeHistogramBuckets int) (*atomic.Int64, bool) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.refs[ref].nanos
+	entry := s.refs[ref]
+	return entry.nanos, entry.isNativeHistogram != hasNativeHistograms || entry.numNativeHistogramBuckets != numNativeHistogramBuckets
 }
 
-func (s *seriesStripe) findOrCreateEntryForSeries(ref uint64, series labels.Labels, nowNanos int64) (*atomic.Int64, bool) {
+func (s *seriesStripe) findAndUpdateOrCreateEntryForSeries(ref uint64, series labels.Labels, nowNanos int64, hasNativeHistograms bool, numNativeHistogramBuckets int) (*atomic.Int64, bool) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	matches := s.matchers.matches(series)
+	matchesLen := matches.len()
+
 	// Check if already exists within the entries.
 	// This repeats findEntryForSeries(), but under write lock.
 	entry, ok := s.refs[ref]
 	if ok {
+		if entry.isNativeHistogram && !hasNativeHistograms {
+			s.activeNativeHistograms--
+			s.activeNativeHistogramBuckets -= entry.numNativeHistogramBuckets
+			for i := 0; i < matchesLen; i++ {
+				match := matches.get(i)
+				s.activeMatchingNativeHistograms[match]--
+				s.activeMatchingNativeHistogramBuckets[match] -= entry.numNativeHistogramBuckets
+			}
+			entry.isNativeHistogram = false
+			entry.numNativeHistogramBuckets = 0
+			s.refs[ref] = entry
+		} else if !entry.isNativeHistogram && hasNativeHistograms {
+			s.activeNativeHistograms++
+			s.activeNativeHistogramBuckets += numNativeHistogramBuckets
+			for i := 0; i < matchesLen; i++ {
+				match := matches.get(i)
+				s.activeMatchingNativeHistograms[match]++
+				s.activeMatchingNativeHistogramBuckets[match] += numNativeHistogramBuckets
+			}
+			entry.isNativeHistogram = true
+			entry.numNativeHistogramBuckets = numNativeHistogramBuckets
+			s.refs[ref] = entry
+		} else if entry.numNativeHistogramBuckets != numNativeHistogramBuckets {
+			diff := numNativeHistogramBuckets - entry.numNativeHistogramBuckets
+			s.activeNativeHistogramBuckets += diff
+			for i := 0; i < matchesLen; i++ {
+				s.activeMatchingNativeHistogramBuckets[matches.get(i)] += diff
+			}
```

Comment:
Note: I've tried to come up with a way to make this go away, but couldn't. I guess it's fine since usually the `matchesLen` will be 1 or a small number as you'd one to do charge-back to one time only.